src/game/logic/object/update/physicsupdate.cpp

```diff
@@ -648,3 +912,504 @@ Module *PhysicsBehavior::Friend_New_Module_Instance(Thing *thing, ModuleData con
+    return obj != nullptr && obj->Get_ID() == m_previousOverlap;
+}
+
+bool PhysicsBehavior::Is_Ignoring_Collisions_With(ObjectID id) const
+{
+    return id != OBJECT_UNK && id == m_ignoreCollisionsWith;
+}
+
+bool PhysicsBehavior::Is_Currently_Overlapped(Object *obj) const
+{
+    return obj != nullptr && obj->Get_ID() == m_currentOverlap;
+}
+
+void PhysicsBehavior::Test_Stunned_Unit_For_Destruction()
+{
+    if (Get_Flag(STUNNED)) {
+        Object *obj = Get_Object();
+        const Coord3D *pos = obj->Get_Position();
+        AIUpdateInterface *update = obj->Get_AI_Update_Interface();
+
+        if (obj->Get_Transform_Matrix()->Get_Z_Vector().Z < 0.0f || obj->Is_Outside_Map()
+            || (update != nullptr)
+                && (g_theTerrainLogic->Is_Cliff_Cell(pos->x, pos->y)
+                        && !update->Has_Locomotor_For_Surface(LOCOMOTOR_SURFACE_CLIFF)
+                    || g_theTerrainLogic->Is_Underwater(pos->x, pos->y, nullptr, nullptr)
+                        && !update->Has_Locomotor_For_Surface(LOCOMOTOR_SURFACE_WATER))) {
+            obj->Kill(DAMAGE_UNRESISTABLE, DEATH_NORMAL);
+        }
+    }
+}
+
+void PhysicsBehavior::Apply_YPR_Damping(float damping)
+{
+    m_pitchRate = damping * m_pitchRate;
+    m_rollRate = damping * m_rollRate;
+    m_yawRate = damping * m_yawRate;
+    if (m_pitchRate == 0.0f && m_rollRate == 0.0f && m_yawRate == 0.0f) {
+        Set_Flag(MOVING, false);
+    } else {
+        Set_Flag(MOVING, true);
+    }
+}
+
+float PhysicsBehavior::Get_Aerodynamic_Friction() const
+{
+    float friction = Get_Physics_Behavior_Module_Data()->m_aerodynamicCoeff + m_extraFriction;
+
+    if (friction < 0.0f) {
+        friction = 0.0f;
+    }
+
+    if (friction > 0.99f) {
+        friction = 0.99f;
+    }
+
+    return friction;
+}
+
+float PhysicsBehavior::Get_Forward_Friction() const
+{
+    float friction = Get_Physics_Behavior_Module_Data()->m_forwardFriction + m_extraFriction;
+
+    if (friction < 0.0f) {
+        friction = 0.0f;
+    }
+
+    if (friction > 0.99f) {
+        friction = 0.99f;
+    }
+
+    return friction;
+}
+
+float PhysicsBehavior::Get_Lateral_Friction() const
+{
+    float friction = Get_Physics_Behavior_Module_Data()->m_lateralFriction + m_extraFriction;
+
+    if (friction < 0.0f) {
+        friction = 0.0f;
+    }
+
+    if (friction > 0.99f) {
+        friction = 0.99f;
+    }
+
+    return friction;
+}
+
+float PhysicsBehavior::Get_Z_Friction() const
+{
+    float friction = Get_Physics_Behavior_Module_Data()->m_zFriction + m_extraFriction;
+
+    if (friction < 0.0f) {
+        friction = 0.0f;
+    }
+
+    if (friction > 0.99f) {
+        friction = 0.99f;
+    }
+
+    return friction;
+}
+
+void PhysicsBehavior::Add_Velocity_To(const Coord3D *velocity)
+{
+    if (velocity != nullptr) {
+        m_vel.Add(velocity);
+    }
+}
+
+void PhysicsBehavior::Scrub_Velocity_Z(float desired_velocity)
+{
+    if (GameMath::Fabs(desired_velocity) < 0.001000000047497451f) {
```

Comment:
is 0.001000000047497451f not just identical to 0.001f ?

Reply:
fixed.